Firestore/core/src/firebase/firestore/remote/buffered_writer.cc

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/buffered_writer.h"
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+void BufferedWriter::DiscardUnstartedWrites() {
+  queue_ = {};
+}
+
+void BufferedWriter::EnqueueWrite(std::unique_ptr<GrpcOperation> write) {
+  HARD_ASSERT(write, "Trying to enqueue a null write operation");
+  queue_.push(std::move(write));
+  TryStartWrite();
+}
+
+void BufferedWriter::TryStartWrite() {
+  if (empty()) {
+    return;
+  }
+  if (has_active_write_) {
+    return;
+  }
+
+  has_active_write_ = true;
+  // Once an operation is executed, the ownership is (implicitly) transferred to
+  // the completion queue.
+  GrpcOperation* write_operation = queue_.front().release();
```

Comment:
As written the tests leak this `write_operation` which is something we should avoid.

Additionally, `EnqueueWrite` takes a `std::unique_ptr` which makes it seem as if it's going to conform to simple modern C++ ownership principles, but this does anything but that making this a trap for the unwary.

`std::shared_ptr` would be marginally better but it seems as the operation would need to be able to give a reference to itself, which while possible, points to the possibility that we may be misdesigning this, since the buffers themselves are already reference counted and all we really want to do is hand those off.

If `GrpcOperation` were a value type containing the buffer and the tag, `BufferedWriter` could use `AsyncWriterInterface` as its delegate and then we wouldn't need any complicated hand-off ownership semantics.

Let's discuss this morning in the context of the larger PR and see if that makes sense.

Reply:
Update: tried out manager approach in the follow-up PR -- see [this commit](https://github.com/firebase/firebase-ios-sdk/pull/1706/commits/7240355a38a759dc2e31b7962e6eb13cda1eeef4). Let me know what you think.